### PR TITLE
[Access] Added metrics for transaction validator

### PR DESCRIPTION
**Pull request by Guitarheroua**
#6128

Introduced metrics implementation for transaction validator

DEPENDS ON  https://github.com/onflow/flow-emulator/pull/723

### CODE REVIEW

File: access/validator.go
Comment (peterargue) [inline access/validator.go:424]: I think the cast was needed. `DecodeVerifyPayerBalanceResult` returns a `cadence.Bool` not a native `bool`

File: module/metrics/collection.go
Comment (peterargue) [inline module/metrics/collection.go:21]: ❤️

File: module/metrics/transaction_validation.go
Comment (peterargue) [inline module/metrics/transaction_validation.go:27]: we won't be able to use `txID` as a label for these. Since the value changes for every tx, this would create a new counter time series for every transaction, resulting in millions of new metrics per day. Labels can only be used when there's a small fixed set of possible values. `reason` below is good though.

File: access/validator.go
Comment (peterargue) [inline access/validator.go:174]: can you make these string values consts

File: access/validator.go
Comment (peterargue) [inline access/validator.go:160]: can you make this a member of `TransactionValidator`. I think it will make the code a little more explicit

File: module/metrics/transaction_validation.go
Comment (peterargue) [inline module/metrics/transaction_validation.go:23]: the metric is tracking successful validations. I think using `total` here will give people the impression that it's the total validation attempts.

Can you rename this to something like `transaction_validation_succeeded`

**Review by peterargue — COMMENTED** [review]
added a couple small comments, but otherwise this looks good

File: access/validator.go
Comment (jordanschalm) [inline access/validator.go:26]: We store other metric label constants in https://github.com/onflow/flow-go/blob/master/module/metrics/labels.go - for consistency probably best to put these labels there as well.

File: module/metrics/transaction_validation.go
Comment (jordanschalm) [inline module/metrics/transaction_validation.go:27]: Let's use the same namespace here regardless of the node being constructed (`access` is fine). The same metric data is being produced, regardless of whether transaction validation is running on a Collection or Consensus node, so we should use the same qualified metric name.

If we want to see transaction validation metrics from a specific node role, labels are the best tool to do that, and we already have `role` labels defined for all our metrics that allow us to do this. 

Namespaces help to produce consistently structured fully qualified metric names, but they shouldn't strictly correspond to node roles if a metric is used across multiple roles. For example, the finalized height metric uses the namespace `consensus`, but all node roles produce this metric (with the same qualified name), because all nodes follow consensus.

File: module/metrics/transaction_validation.go
Comment (jordanschalm) [inline module/metrics/transaction_validation.go:39]:
```suggestion
		transactionValidated: promauto.NewCounter(prometheus.CounterOpts{
			Name:      "transaction_validation_succeeded",
			Namespace: namespaceAccess,
			Subsystem: subsystemTransactionValidation,
			Help:      "counter for the validated transactions",
		}),
		transactionValidationFailed: promauto.NewCounterVec(prometheus.CounterOpts{
			Name:      "transaction_validation_failed",
			Namespace: namespaceAccess,
			Subsystem: subsystemTransactionValidation,
			Help:      "counter for the failed transactions validation",
```

File: module/metrics/transaction_validation.go
Comment (jordanschalm) [reply module/metrics/transaction_validation.go]: The `total` suffix is part of Prometheus [best practices](https://prometheus.io/docs/practices/naming/#metric-names) for counters (guessing that's why it was there to begin with). 

Something like `transaction_validation_successes_total` would communicate both that it's a counter and that it's counting successes (not validations in general, like Peter pointed out).

File: module/metrics/transaction_validation.go
Comment (jordanschalm) [inline module/metrics/transaction_validation.go:15]:
```suggestion
```

**Review by jordanschalm — APPROVED** [review]
🎸

File: access/validator.go
Comment (peterargue) [inline access/validator.go:195]: it occurred to me during our chat this morning that we need a metric to track skipped validations. right now those will all appear as valid and we'll have no way to identify failures due to errors.

Can you add a 3rd metric to track skipped validations

File: module/metrics/transaction_validation.go
Comment (peterargue) [inline module/metrics/transaction_validation.go:32]: circling back to Jordan's comment about the counter best practice. let's add total
```suggestion
			Name:      "transaction_validation_skipped_total",
```

File: module/metrics/transaction_validation.go
Comment (peterargue) [inline module/metrics/transaction_validation.go:38]:
```suggestion
			Name:      "transaction_validation_failed_total",
```

File: access/validator.go
Comment (peterargue) [inline access/validator.go:198]: don't forget to call the new skipped metric here

**Review by peterargue — COMMENTED** [review]
looks good. I'll approve and merge after the emulator change is merged and the replace is removed from this PR

**Comment by peterargue** [comment]
@Guitarheroua please fix conflicts

**Review by peterargue — APPROVED** [review]

**Pull request merged**
